### PR TITLE
[skrifa] tthint: low hanging fruit

### PR DESCRIPTION
**Pull request by dfrg**
Implements some of the simpler outline instructions that deal with point flags and also a couple from the "reading and writing data" group.

Includes a small change to the definition code: move the remap of InvalidDefinition->UnhandledOpcode error to actual definition lookup.

### CODE REVIEW

File: skrifa/src/outline/glyf/hint/engine/data.rs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    /// instruction stream. It makes it possible to treat point sizes below or
+    /// above a certain threshold differently.
+    ///
+    /// See <https://learn.microsoft.com/en-us/typography/opentype/spec/tt_instructions#measure-point-size>
+    /// and <https://gitlab.freedesktop.org/freetype/freetype/-/blob/57617782464411201ce7bbc93b086c1b4d7d84a5/src/truetype/ttinterp.c#L2623>
+    pub(super) fn op_mps(&mut self) -> OpResult {
+        // Note: FreeType computes this at
+        // <https://gitlab.freedesktop.org/freetype/freetype/-/blob/57617782464411201ce7bbc93b086c1b4d7d84a5/src/truetype/ttdriver.c#L392>
+        // which is mul_div(ppem, 64 * 72, resolution) where resolution
+        // is always 72 for our purposes (Skia), resulting in ppem * 64.
```
Comment (rsheeter) [inline skrifa/src/outline/glyf/hint/engine/data.rs:43]: Isn't this potentially problemmatic for a non-Skia client? Would it make sense to store resolution, default 72, and support the mul_div?

File: skrifa/src/outline/glyf/hint/engine/outline.rs [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+    /// the contour describing part of a glyph outline is redefined.
+    ///
+    /// See <https://learn.microsoft.com/en-us/typography/opentype/spec/tt_instructions#flip-point>
+    /// and <https://gitlab.freedesktop.org/freetype/freetype/-/blob/57617782464411201ce7bbc93b086c1b4d7d84a5/src/truetype/ttinterp.c#L5002>
+    pub(super) fn op_flippt(&mut self) -> OpResult {
+        let count = self.graphics_state.loop_counter as usize;
+        self.graphics_state.loop_counter = 1;
+        // In backward compatibility mode, don't flip points after IUP has
+        // been done.
+        if self.graphics_state.backward_compatibility
```
Comment (rsheeter) [inline skrifa/src/outline/glyf/hint/engine/outline.rs:33]: In this context it's not entirely obvious what backward_compatibility would be with, I wonder if we should name it to more clearly indicate? For example, `if self.graphics_state.freetype_compat` makes it very clear what we're compatible with.

File: skrifa/src/outline/glyf/hint/engine/outline.rs [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+            (false, false) => None,
+        };
+        if let Some(coord_axis) = coord_axis {
+            self.graphics_state.zp0_mut().untouch(p, coord_axis)?;
+        }
+        Ok(())
+    }
+
+    /// Helper for FLIPRGON and FLIPRGOFF.
+    fn do_fliprg(&mut self, on: bool) -> OpResult {
```
Comment (rsheeter) [inline skrifa/src/outline/glyf/hint/engine/outline.rs:118]: `do_fliprg(true/false)` is pretty obtuse, an enum or a long name wouldn't hurt. Mildly surprisingly it doesn't seem to *flip* but rather to set so perhaps something like:

```rust
    /// Helper for FLIPRGON and FLIPRGOFF.
    fn set_on_curve_for_range(&mut self, on: bool) -> OpResult {
```

File: skrifa/src/outline/glyf/hint/engine/outline.rs [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+            return Ok(());
+        }
+        self.graphics_state
+            .zone_mut(ZonePointer::Glyph)
+            .set_on_curve(low_point, high_point, on)
+    }
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (rsheeter) [inline skrifa/src/outline/glyf/hint/engine/outline.rs:138]: Missing tests for nop when back compat is set and did_iup is true?

File: skrifa/src/outline/glyf/hint/engine/outline.rs [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+        // high_point is inclusive but Zone::set_on_curve takes an exclusive
+        // range
+        let high_point = self.value_stack.pop_usize()? + 1;
+        let low_point = self.value_stack.pop_usize()?;
+        // In backward compatibility mode, don't flip points after IUP has
+        // been done.
+        if self.graphics_state.backward_compatibility
+            && self.graphics_state.did_iup_x
+            && self.graphics_state.did_iup_y
+        {
```
Comment (rsheeter) [inline skrifa/src/outline/glyf/hint/engine/outline.rs:128]: is did_iup for only one of x and y a state we should actually encounter? If so, should we or should we not modify things?

File: skrifa/src/outline/glyf/hint/graphics_state/mod.rs [hunk 6]
```diff
@@ -104,6 +104,10 @@ pub struct GraphicsState<'a> {
     ///
```
Comment (rsheeter) [inline skrifa/src/outline/glyf/hint/graphics_state/mod.rs:104]: `backward_compatibility` is a bit hard to follow. Despite the comment and the freetype link I'm still not entirely clear what we're being backward compatible with. If possible expand the comment and/or rename the field.

File: skrifa/src/outline/glyf/hint/graphics_state/mod.rs [hunk 7]
```diff
@@ -104,6 +104,10 @@ pub struct GraphicsState<'a> {
...
     /// See <https://gitlab.freedesktop.org/freetype/freetype/-/blob/57617782464411201ce7bbc93b086c1b4d7d84a5/src/truetype/ttinterp.h#L195>
     pub is_pedantic: bool,
+    /// Set to true when IUP has been executed in the horizontal direction.
+    pub did_iup_x: bool,
+    /// Set to true when IUP has been executed in the vertical direction.
+    pub did_iup_y: bool,
```
Comment (rsheeter) [inline skrifa/src/outline/glyf/hint/graphics_state/mod.rs:110]: Do we typically iup in both directions at the same time or is there a period when instructions could run when only one of these are set? - we have if statements for `back compat && did_iup_x && did_iup_y` and it's not clear to me if only one of them being set is a realistic state.

Basically why can't we have just `did_iup: bool,`

**Review by rsheeter — APPROVED** [review]
Couple of small questions but generally lgtm

File: skrifa/src/outline/glyf/hint/engine/outline.rs (hunk 2 above)
Comment (dfrg) [reply skrifa/src/outline/glyf/hint/engine/outline.rs:33]: This isn’t a FT compatibility thing. It’s a feature of the MS rasterizer that allows fonts (or specific glyphs) to opt in to cleartype native mode. Basically, both MS and FT implement a bunch of hacks for non cleartype fonts that prevent bytecode from modifying outlines in certain ways. This flag (which takes its name from FT) controls that behavior and is toggled by the INSTCTRL instruction.

File: skrifa/src/outline/glyf/hint/engine/outline.rs (hunk 4 above)
Comment (dfrg) [reply skrifa/src/outline/glyf/hint/engine/outline.rs:138]: Oops, yes. Will add them

File: skrifa/src/outline/glyf/hint/engine/outline.rs (hunk 2 above)
Comment (rsheeter) [reply skrifa/src/outline/glyf/hint/engine/outline.rs:33]: Thank you, that's helpful. I suggest putting some of that on the comment for the backward compatibility flag.

File: skrifa/src/outline/glyf/hint/engine/outline.rs (hunk 5 above)
Comment (dfrg) [reply skrifa/src/outline/glyf/hint/engine/outline.rs:128]: Yes, IUP can be done in only one direction but IIRC, it can be faked in the x direction. See note about hacks above :)

File: skrifa/src/outline/glyf/hint/graphics_state/mod.rs (hunk 6 above)
Comment (dfrg) [reply skrifa/src/outline/glyf/hint/graphics_state/mod.rs:104]: I’ll add an extended description based on my explanation above. I think there’s a useful MS resource about this as well. I’ll try to find a link.

File: skrifa/src/outline/glyf/hint/graphics_state/mod.rs (hunk 7 above)
Comment (dfrg) [reply skrifa/src/outline/glyf/hint/graphics_state/mod.rs:110]: Yes, IUP in x and y are different instructions that can be executed separately and modifying the behavior based on this is basically the whole idea behind “modern vertical only hinting” and the backward compatibility flag controls how we do this for fonts that contain bytecode that does full grid fitting.

File: skrifa/src/outline/glyf/hint/engine/data.rs (hunk 1 above)
Comment (dfrg) [reply skrifa/src/outline/glyf/hint/engine/data.rs:43]: To be a bit more concrete, I’d expect a user to ask for a new constructor for `Size` that takes point size and resolution rather than ppem. At that point, we can plumb this through.
